[ticket/14831] Make sure migrations always start with backslash

marc1706 opened the pull request:
Checklist:
- [x] Correct branch: master for new features; 3.2.x, 3.1.x for fixes
- [x] Tests pass
- [x] Code follows coding guidelines: [master / 3.2.x](https://area51.phpbb.com/docs/master/coding-guidelines.html), [3.1.x](https://area51.phpbb.com/docs/31x/coding-guidelines.html)
- [x] Commit follows commit message [format](https://wiki.phpbb.com/Git#Commit_Messages)

Tracker ticket (set the ticket ID to **your ticket ID**):

https://tracker.phpbb.com/browse/PHPBB3-14831

The migration system expects every migration to start with a backslash.
If depends on definitions do not supply that leading backslash, we should
make sure it's added manually before calling the depends on migration.

PHPBB3-14831

michaelcullum commented:
👍

marc1706 commented:
@VSEphpbb updated to be more robust than the previous change. Please try to test this as much as possible ;)

rxu commented on phpBB/phpbb/db/migrator.php:
```diff
@@ -264,10 +277,22 @@ protected function try_apply($name)
+			{
+				if (isset($this->migration_state[preg_replace('#^(?!\\\)#', '\\\$0', $name)]))
+				{
+					$name = preg_replace('#^(?!\\\)#', '\\\$0', $name);
```
To save some ```preg_replace``` calls maybe assign it to some variable before.

rxu commented on phpBB/phpbb/db/migrator.php:
```diff
@@ -264,10 +277,22 @@ protected function try_apply($name)
+				}
+				else if (isset($this->migration_state[preg_replace('#(^\\\)([^\\\].+)#', '$2', $name)]))
+				{
+					$name = preg_replace('#(^\\\)([^\\\].+)#', '$2', $name);
```
Same as above.

VSEphpbb commented:
A little problem encountered with this.

If you already had an extension installed that added a migration dependency without the leading slash to your migrations table. This is what you end up with in the database:
![screen shot 2016-10-23 at 9 01 11 am](https://cloud.githubusercontent.com/assets/303711/19627611/46a42ce0-98ff-11e6-80c3-9f024c0d57c8.png)
You'll notice that the m1 migration appears twice (once without the slash) and all the others depend on the one without the slash.

Then you uninstall that extension (the migration w/o the slash remains in the database). 
![screen shot 2016-10-23 at 9 07 47 am](https://cloud.githubusercontent.com/assets/303711/19627641/3310a81a-9900-11e6-8656-8fc063dc5164.png)

With this PR's changes, if you try to reinstall the extension again, that "rogue" migration is now prevented from running.

I'm not sure if we can do anything about this other than through support and tell people to uninstall the ext, then delete that from their DB?

VSEphpbb commented on phpBB/phpbb/db/migration/data/v31x/migrations_deduplicate_entries.php:
```diff
@@ -0,0 +1,78 @@
+
+namespace phpbb\db\migration\data\v31x;
+
+class migrations_deduplicate_entries extends \phpbb\db\migration\migration
```
remove_duplicate_migration_entries ?

VSEphpbb replied:
or `migrations_remove_duplicates`
or `remove_duplicate_migrations`
:)

marc1706 replied:
I'm guessing you don't like using deduplicate?

VSEphpbb replied:
you are correct sir

VSEphpbb commented on phpBB/phpbb/db/migration/data/v31x/migrations_deduplicate_entries.php:
```diff
@@ -0,0 +1,78 @@
+				$migration_state[$migration['migration_name']] = $migration;
+
+				$migration_state[$migration['migration_name']]['migration_depends_on'] = unserialize($migration['migration_depends_on']);
+				$migration_state[$migration['migration_name']]['migration_data_state'] = !empty($migration['migration_data_state']) ? unserialize($migration['migration_data_state']) : '';
```
Is this even used?

marc1706 replied:
Nope, will remove

VSEphpbb commented:
tested and works

Nicofuma commented on phpBB/phpbb/db/migration/data/v31x/migrations_deduplicate_entries.php:
```diff
@@ -0,0 +1,78 @@
+		foreach ($migration_state as $name => $migration)
+		{
+			$prepended_name = preg_replace('#^(?!\\\)#', '\\\$0', $name);
+			$prefixless_name = preg_replace('#(^\\\)([^\\\].+)#', '$2', $name);
```
I'd prefer `$name = ($name[0] === '\' ? '' : '\').$name;` I find it a lot more readable

Nicofuma commented:
What do we have in database before this PR? Duplicated migrations or just migrations without the leading slash?

JoshyPHP commented on phpBB/phpbb/db/migration/data/v31x/migrations_deduplicate_entries.php:
```diff
@@ -0,0 +1,78 @@
+
+			if ($prepended_name != $name && isset($migration_state[$prepended_name]) && $migration_state[$prepended_name]['migration_depends_on'] == $migration_state[$name]['migration_depends_on'])
+			{
+				$duplicate_migrations[] = $name;
```
I don't know what this code block really does so my apologies if I'm way off base but it looks like it should be `$duplicate_migrations[] = $prepended_name;`

VSEphpbb replied:
Nah it should be $name.

VSEphpbb commented:
@Nicofuma It's possible the DB can have duplicate migrations that are missing a slash.. See the 1st image attachment above.

Nicofuma commented:
I mean, when they is a a migration with a missing slash it is always a duplicate?

marc1706 commented:
Probably, but I don't really want to rely on that which is why I'm checking for both possibilities. Also, if you're having the same migration in the db twice, but with different depends_on() (which really shouldn't happen) it will now not remove those.

Nicofuma commented:
I am asking that because the current migration only remove duplicate, it does not update existig migrations

VSEphpbb commented:
This should only remove duplicates. They don't need updating.

marc1706 commented:
Updated.

hubaishan commented:
This commit is buggy and cause the error explain here https://tracker.phpbb.com/browse/PHPBB3-14935

hubaishan commented:
to correct the bug the function get_valid_name must edited to

`	protected function get_valid_name($name)
	{
		// Try falling back to a valid migration name with or without leading backslash
		if (!isset($this->migration_state[$name]))
		{
			$prepended_name = ($name[0] == '\\' ? '' : '\\') . $name;
			$prefixless_name = $name[0] == '\\' ? substr($name, 1) : $name;

			if (isset($this->migration_state[$prefixless_name]))
			{
				$name = $prefixless_name;
			}
			else 
			{
				$name = $prepended_name;
			}
		}

		return $name;
	}
`

marc1706 commented:
Have you actually tried this and ran the migration? That was the initial bugfix but instead we added a migration that gets rid of the duplicate entries. The migrator was also modified to always find a depends_on() migration with or without leading backslash in order to prevent future duplicates. In the end you should make sure to create a valid migrations file and any extension submitted to the customisations database will be checked for valid definitions both by EPV and the Extensions Team.

hubaishan commented:
Yes I tried it.
the mistake will occur if $name doesn't started with backslash and also it isn't exit in migration table in this case get_valid_name will return the same in $name which doesn't started with backslash.

this will cause error only if a migration (b) depends on migration (a) not started with backslash so try_apply() will check if (a) is exist in migration table, in case (a) is before (b) in (a-z) order function  get_valid_name will give the name which in the migration table (prepended_name), but if  (a) is after (b) in (a-z) order then (a) will not be in the migration table so get_valid_name will return original (a) name (prefixless_name) and try_apply() will call try_apply(a) so (a) will inserted in migration table without pre slash, so system will go in non stop loop, and hanged. 

by editing  get_valid_name with my suggest it will return  $prepended_name as default if it is not exist in migration table without backslash.
